File: app/src/main/java/com/scottquach/homeworkchatbotassistant/adapters/RecyclerChatAdapter.java
package com.scottquach.homeworkchatbotassistant.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.scottquach.homeworkchatbotassistant.MessageType;
import com.scottquach.homeworkchatbotassistant.R;
import com.scottquach.homeworkchatbotassistant.models.MessageModel;

import java.util.List;

import timber.log.Timber;

/**
 * Created by Scott Quach on 9/10/2017.
 */

public class RecyclerChatAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<MessageModel> messageModels;
    private Context context;

    public RecyclerChatAdapter(List<MessageModel> messageModels,Context context) {
        this.context = context;
        this.messageModels = messageModels;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case MessageType.SENT:
                View view = LayoutInflater.from(context).inflate(R.layout.row_message_sent, parent, false);
                return new SentViewHolder(view);
            case MessageType.RECEIVED:
                View view1 = LayoutInflater.from(context).inflate(R.layout.row_message_received, parent, false);
                return new ReceivedViewHolder(view1);
            default:
                View view2 = LayoutInflater.from(context).inflate(R.layout.row_message_sent, parent, false);
                return new SentViewHolder(view2);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case MessageType.SENT:
                ((SentViewHolder) holder).messageViewSent.setText(messageModels.get(position).getMessage());
                break;
            case MessageType.RECEIVED:
                ((ReceivedViewHolder) holder).messageViewReceived.setText(messageModels.get(position).getMessage());
                break;
        }
    }

    @Override
    public int getItemViewType(int position) {
        switch ((int) messageModels.get(position).getType()) {
            case MessageType.SENT:
                return MessageType.SENT;
            case MessageType.RECEIVED:
                return MessageType.RECEIVED;
            default:
                Timber.d("couldn't get view type, resorting to default");
                return MessageType.SENT;
        }
    }

    @Override
    public int getItemCount() {
        return messageModels.size();
    }

    public void addMessage(MessageModel model) {
        this.messageModels.add(model);
        notifyItemInserted(getItemCount() - 1);
    }

    public class SentViewHolder extends RecyclerView.ViewHolder{

        public TextView messageViewSent;

        public SentViewHolder(View itemView) {
            super(itemView);
            messageViewSent = itemView.findViewById(R.id.view_message_sent);
        }
    }



    public class ReceivedViewHolder extends RecyclerView.ViewHolder {

        public TextView messageViewReceived;

        public ReceivedViewHolder(View itemView) {
            super(itemView);
            messageViewReceived = itemView.findViewById(R.id.view_message_received);
        }
    }
}

File: app/src/main/java/com/scottquach/homeworkchatbotassistant/adapters/RecyclerChatAdapter.kt
package com.scottquach.homeworkchatbotassistant.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.scottquach.homeworkchatbotassistant.MessageType
import com.scottquach.homeworkchatbotassistant.R
import com.scottquach.homeworkchatbotassistant.models.MessageModel

import timber.log.Timber

/**
 * Created by Scott Quach on 9/10/2017.
 */

class RecyclerChatAdapter(private val context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val userMessages = mutableListOf<MessageModel>()


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            MessageType.SENT -> {
                val view = LayoutInflater.from(context).inflate(R.layout.row_message_sent, parent, false)
                return SentViewHolder(view)
            }
            MessageType.RECEIVED -> {
                val view1 = LayoutInflater.from(context).inflate(R.layout.row_message_received, parent, false)
                return ReceivedViewHolder(view1)
            }
            else -> {
                val view2 = LayoutInflater.from(context).inflate(R.layout.row_message_sent, parent, false)
                return SentViewHolder(view2)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            MessageType.SENT -> (holder as SentViewHolder).messageViewSent.text = userMessages[position].message
            MessageType.RECEIVED -> (holder as ReceivedViewHolder).messageViewReceived.text = userMessages[position].message
        }
    }

    override fun getItemViewType(position: Int): Int {
        when (userMessages[position].type.toInt()) {
            MessageType.SENT -> return MessageType.SENT
            MessageType.RECEIVED -> return MessageType.RECEIVED
            else -> {
                Timber.d("couldn't get view type, resorting to default")
                return MessageType.SENT
            }
        }
    }

    override fun getItemCount(): Int {
        return userMessages.size
    }

    fun addMessage(model: MessageModel) {
        userMessages.add(model)
        notifyItemInserted(itemCount)
    }

    fun addData(newData: List<MessageModel>) {
        for (model in newData) {
            userMessages.add(model)
        }
    }

    inner class SentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var messageViewSent: TextView

        init {
            messageViewSent = itemView.findViewById(R.id.view_message_sent)
        }
    }


    inner class ReceivedViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var messageViewReceived: TextView

        init {
            messageViewReceived = itemView.findViewById(R.id.view_message_received)
        }
    }
}
